streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -108,6 +124,44 @@ private void unlock() {
         version.topologyLock.unlock();
     }
 
+    public Collection<String> sourceTopicsForTopology(final String name) {
+        return builders.get(name).sourceTopicCollection();
+    }
+
+    public boolean needsUpdate(final String threadName) {
+        return threadVersions.get(threadName) < topologyVersion();
+    }
+
+    public void registerThread(final String threadName) {
+        threadVersions.put(threadName, 0L);
+    }
+
+    public void unregisterThread(final String threadName) {
+        threadVersions.remove(threadName);
+    }
+
+    public void reachedLatestVersion(final String threadName) {
+        try {
+            lock();
+            final Iterator<TopologyVersionWaiters> iterator = version.activeTopologyWaiters.listIterator();
```

Comment:
Correct me if I think crazily here :P My read is that the topology version of thread can ONLY be incremented (maybe by more than 1), but never decremented. So we actually do not need to remember what are the "waiting" versions as a list, instead we can just keep a list of futures for each thread, and when a thread has successfully `handleTopologyUpdates`, we can immediately complete all the currently maintained futures since there will be no futures that can ever be related to a newer version that this thread has just updated to.

With that, we can 1) get rid of the `topologyVersionWaiters`, 2) do not need the nest loop of while + stream() below fo complete futures. Instead we just keep a list of futures without any additional futures associated with them per thread.

Reply:
I reason I would prefer to keep it as a list of `topologyVersionWaiters` is because while threads are reaching a version the topology can be updated again. Maybe I have been looking at this too long but it seems easier to keep track of this way to me. We have a complexity of the number of times the topology have updated * the number of threads. Realistically the number of threads are bounded so we can treat as a constant. And we would have the same number of futures (times the topology has been updated) to complete either way. I think the little extra verbosity makes it easier to really understand what is going on. But then maybe I am too close to this code :|

Reply:
I see @guozhangwang 's point, but I think it's not worth obsessing over. @wcarlson5 how about you just file a V1.1 ticket for cleanup like this? Doesn't seem worth blocking the PR over :)